[GH-105] Add ALP dataset for floating point columns

prtkgaur opened the pull request:
Add dataset needed for testing performance of floating point values.

- related to  https://github.com/apache/parquet-format/issues/533

- Closes https://github.com/apache/parquet-testing/issues/105

sfc-gh-pgaur committed:
Add dataset for floating point columns

sfc-gh-pgaur committed:
List names of all the columns

prtkgaur commented:
> Add dataset needed for testing performance of floating point values.
Files are in csv format, I did this for easier testing. Let me know if the you (the reviewer thinks otherwise).

emkornfield commented:
Thanks for the PR:
1.  Lets finalize on the spec changes first (I think there is still some easy room for reducing size).
2. Lets at least have two simple examples, where the content is well defined that covers the following cases:
   a. Null values
   b. 0 Exceptions for blocks
   c. 1 exception for a block
   d. >1 exception for a block.
   e. A block with all exceptions.
   f. differning exponent/value per block.

(are there other data variations that are important to cover?)

sfc-gh-pgaur committed:
Add ALP-encoded parquet test files and expected CSVs

Generated from floatingpoint_spotify1.csv (9 double columns, 15000 rows)
and floatingpoint_arade.csv (4 double columns, 15000 rows) using the C++
generate_alp_parquet utility with ALP encoding, no dictionary, uncompressed.

These files are used for cross-language ALP correctness testing: read the
parquet file and verify all values match the expected CSV bit-exactly.

sfc-gh-pgaur committed:
Add Java-generated ALP-encoded parquet test files

sfc-gh-pgaur committed:
Add float32 ALP-encoded test data files

Add C++ and Java-generated float32 ALP parquet files and expected CSVs
for cross-language interop testing of ALP encoding with FLOAT type.

alamb commented:
Checking it out...

alamb commented:
Sorry I didn't post this earlier. My biggest concern with this PR is that it adds many bytes of example data. Since this repo is checked out many times by many CI jobs (e.g. arrow, arrow-rs, etc) I think it is important to keep the size down

Furthermore, I am not sure that this covers the major parameter (I need to review the files carefully first)

I will file a ticket that records what I think is needed, and link this PR

alamb commented:
I also filed a ticket to track this issue and explain what I think is needed in the example files
- https://github.com/apache/parquet-testing/issues/105

vinooganesh commented:
Hey all, quick heads up, I just opened a stacked PR (prtkgaur/parquet-testing#1) with six Java-written ALP fixtures from apache/parquet-java#3397 to broaden the cross-language test surface here.

Four of them are dataset-derived (~960 KB combined) and cover both page versions (V1/V2) × both ALP vector sizes (1024/4096) for float and double, using the existing arade and spotify1 source data. The other two are a small corner-case fixture (~60 KB) targeting the design list in #105 — vectors with no exceptions, one exception, all-NaN, NaN+Inf+(-0.0), constant (bit_width=0), differing exponents per vector, and optional columns with nulls, both f32 and f64 — plus a sidecar _expect.csv ground truth emitted directly from the construction recipe, so the corner-case file is independently verifiable without rerunning the Java generator.

I've already run Arrow C++ ALP from apache/arrow#48345 against all six and bit-compared every value against the _expect.csv truth — 342K values, 0 mismatches. The full local matrix (17 fixtures across all V1/V2 × 1024/4096 combinations) comes out to 1.59M values verified, also 0 mismatches.
  
 On size: @alamb raised concerns in #105 about test data bloating CI checkouts, so I deliberately kept this PR small — ~1.1 MB  total. I have 11 more fixtures locally from the full matrix (~2.5 MB more) if anyone wants additional coverage, but figured the four representative variants plus the corner cases were enough to verify the axes work without growing the repo unnecessarily.

Happy to add or trim, let me know what would be most useful.

alamb committed:
Merge remote-tracking branch 'origin/master' into alpFloatingPointDataset

alamb committed:
Document what is in the example ALP files

alamb reviewed (commented):
Thank you @prtkgaur and @vinooganesh 

I think the data files on this branch make a lot of sense as a way to test the first two implementations. However, I don't think we can merge this PR as is:

# Lack of missing / interesting values:
1. There doesn't seem to be any files with nan / inf type values (this is partly addressed in https://github.com/prtkgaur/parquet-testing/pull/1)

# File Size
This PR (not including the one from https://github.com/prtkgaur/parquet-testing/pull/1) 
increases the size of the parquet-testing directory by 3.8x (3.7M --> 14M)


Specifically in this branch:

```shell
$ du -s -h /tmp/master/data/
3.7M	/tmp/master/data/
```
On master
```shell
$ du -s -h data/
 14M	data/
```

# Path forward
In this repo I think we need a minimal exemplar dataset. Specifically I would suggest
a  single parquet file with multiple columns (to keep the size down) and columns with the same values encoded with multiple encodings (so downstream decoders can compare with known results)

I spent some time sketching out what I think would be needed in https://github.com/apache/parquet-testing/issues/105#issuecomment-5172111570 and will try and make such a file tomorrow

alamb commented:
Superceded by https://github.com/apache/parquet-testing/pull/119 so closing this one for now
